### ISSUE TITLE
ctrl-c during open terminates newsboat

### ISSUE THREAD

**Issue by allanwind** [newsboat/newsboat#2202]
**Newsboat version (copy the output of `newsboat -v` or the first line of `git show`)**: r2.28-51-g33117-dirty

Config file (copy from ~/.newsboat/config or ~/.config/newsboat/config):

```
article-sort-order title
articlelist-title-format "%U"
bind-key J next-feed
bind-key K prev-feed
bind-key j next
bind-key k prev
bind-key o open-in-browser-and-mark-read articlelist
confirm-mark-feed-read no
datetime-format "%F %T"
error-log ~/.local/share/newsboat/error.log
feedlist-format "%4i %n %11u %L - %t"
ignore-mode display
keep-articles-days 365
prepopulate-query-feeds yes
reload-threads 32
run-on-startup reload-all;open
show-read-articles no
show-read-feeds no
unbind-key C

```

Steps to reproduce the issue:

1. I upgraded Firefox today but have yet to restart the process.  Firefox is in a funky state where it just hangs when you try to open a new url.
2. To interrupt the hanging external process (xdg-open), I press ctrl-c, which terminates the open as expected, but also the newboat process itself.

**Other info you think is relevant**:

It's pretty esoteric but might be worth figuring out why signal handling doesn't work.

**Comment by Minoru** [comment]
Hi, sorry for the delayed response!

I compiled the same commit (331176fdd9fdbd876f37b33f22a1a7054a13b2f0) and ran Newsboat with the following config:

```
browser data/2202-browser.sh
```

data/2202-browser.sh contains:

```sh
#!/bin/sh

while true
do
    sleep 1 || break
done
```

Then I started Newsboat, opened an article, and pressed `o` to open the URL. Newsboat started the script, and then I pressed <kbd>^C</kbd> This returned me to Newsboat, which displayed "Failed to spawn browser" in the status line (at the bottom of the screen).

In short, I can't reproduce this.

It's possible that my script doesn't emulate the situation well enough. Your config doesn't have `browser` set, so can you please run `echo $BROWSER` from your shell and tell me what it displays?

**Comment by allanwind** [comment]
Thanks for trying.  Is there any specific data you would like me 
to collect if I run into this again?  It was "reproducible" for me 
in the sense that I was surprised by the behavior so I tried it a 
few times.

The variable is not set.  I am guessing that `xdg-open` invokes
`default-web-browser` which is set to `firefox-esr.desktop`
which eventually calls `/usr/lib/firefox-esr/firefox-esr %u`.

**Comment by Minoru** [comment]
For now I'd like to understand what exactly is being run. Your config doesn't contain a `browser` setting, so Newsboat would look up the `BROWSER` environment variable, and since that's not set, Newsboat will use `lynx`. This is clearly **not** what happens on your machine, since you've been able to open links in Firefox. Can you please double-check that you provided your entire config? Another avenue for inquiry would be the `-dirty` tag in the version, which means you have some local modifications to the Git repo — can you show the output of `git diff` please?

**Comment by allanwind** [comment]
It looks like github silently dropped the reply I sent you on "Mon, 3 Oct 2022 16:48:31 -0400" :-(.  Let me try resend it.

**Comment by Minoru** [comment]
> `$ echo $BROSER`

This should say "BRO*W*SER", with a "W". I sure hope this typo is a culprit, because the changes to Cargo.lock don't explain this behaviour, and `utils::get_default_browser()` is trivial and obviously can't return "firefox" unless `BROWSER` is set that way:

https://github.com/newsboat/newsboat/blob/8d7e6ddc35f6e7ed79b9dc38b0115c5a44cceb73/rust/libnewsboat/src/utils.rs#L151-L154

**Comment by allanwind** [comment]
Ack typo but no difference:

```
$ echo $BROWSER
$
```

**Comment by allanwind** [comment]
I provided a copy of my config but omitted ignore-article:

```
$ grep -v ^ignore-article ~/.config/newsboat/config
article-sort-order title
articlelist-title-format "%U"
bind-key J next-feed
bind-key K prev-feed
bind-key j next
bind-key k prev
bind-key o open-in-browser-and-mark-read articlelist
confirm-mark-feed-read no
datetime-format "%F %T"
error-log ~/.local/share/newsboat/error.log
feedlist-format "%4i %n %11u %L - %t"
ignore-mode display
keep-articles-days 365
prepopulate-query-feeds yes
reload-threads 32
run-on-startup reload-all;open
show-read-articles no
show-read-feeds no
unbind-key C
```

and here is the double check you requested:

```
$ grep browser ~/.config/newsboat/config
bind-key o open-in-browser-and-mark-read articlelist
$ echo $BROSER

```

It's dirty because you asked me to update rust/cargo or whatever 
it was:
```
diff --git a/Cargo.lock b/Cargo.lock
index db698b40..5cf2f831 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -26,6 +26,15 @@ dependencies = [
   "memchr",
  ]
  
+[[package]]
+name = "android_system_properties"
+version = "0.1.4"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "d7ed72e1635e121ca3e79420540282af22da58be50de153d36f81ddc6b83aa9e"
+dependencies = [
+ "libc",
+]
+
  [[package]]
  name = "autocfg"
  version = "1.1.0"
@@ -34,9 +43,9 @@ checksum = "d468802bab17cbc0cc575e9b053f41e72aa36bfa6b7f55e3529ffa43161b97fa"
  
  [[package]]
  name = "backtrace"
-version = "0.3.65"
+version = "0.3.66"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "11a17d453482a265fd5f8479f2a3f405566e6ca627837aaddb85af8b1ab8ef61"
+checksum = "cab84319d616cfb654d03394f38ab7e6f0919e181b1b57e1fd15e7fb4077d9a7"
  dependencies = [
   "addr2line",
   "cc",
@@ -49,9 +58,9 @@ dependencies = [
  
  [[package]]
  name = "bit-set"
-version = "0.5.2"
+version = "0.5.3"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "6e11e16035ea35e4e5997b393eacbf6f63983188f7a2ad25bfb13465f5ad59de"
+checksum = "0700ddab506f33b20a03b13996eccd309a48e5ff77d0d95926aa0210fb4e95f1"
  dependencies = [
   "bit-vec",
  ]
@@ -74,6 +83,12 @@ version = "0.1.6"
  source = "registry+https://github.com/rust-lang/crates.io-index"
  checksum = "0d8c1fef690941d3e7788d328517591fecc684c084084702d6ff1641e993699a"
  
+[[package]]
+name = "bumpalo"
+version = "3.10.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "37ccbd214614c6783386c1af30caf03192f17891059cecc394b4fb119e363de3"
+
  [[package]]
  name = "byteorder"
  version = "1.4.3"
@@ -94,14 +109,16 @@ checksum = "baf1de4339761588bc0619e3cbc0120ee582ebb74b53b4efbf79117bd2da40fd"
  
  [[package]]
  name = "chrono"
-version = "0.4.19"
+version = "0.4.22"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "670ad68c9088c2a963aaa298cb369688cf3f9465ce5e2d4ca10e6e0098a1ce73"
+checksum = "bfd4d1b31faaa3a89d7934dbded3111da0d2ef28e3ebccdb4f0179f5929d1ef1"
  dependencies = [
- "libc",
+ "iana-time-zone",
+ "js-sys",
   "num-integer",
   "num-traits",
   "time",
+ "wasm-bindgen",
   "winapi",
  ]
  
@@ -115,6 +132,12 @@ dependencies = [
   "unicode-width",
  ]
  
+[[package]]
+name = "core-foundation-sys"
+version = "0.8.3"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "5827cebf4670468b8772dd191856768aedcb1b0278a04f989f7766351917b9dc"
+
  [[package]]
  name = "curl-sys"
  version = "0.4.56+curl-7.83.1"
@@ -251,9 +274,22 @@ dependencies = [
  
  [[package]]
  name = "gimli"
-version = "0.26.1"
+version = "0.26.2"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "78cc372d058dcf6d5ecd98510e7fbc9e5aec4d21de70f65fea8fecebcd881bd4"
+checksum = "22030e2c5a68ec659fde1e949a745124b48e6fa8b045b7ed5bd1fe4ccc5c4e5d"
+
+[[package]]
+name = "iana-time-zone"
+version = "0.1.44"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "808cf7d67cf4a22adc5be66e75ebdf769b3f2ea032041437a7061f97a63dad4b"
+dependencies = [
+ "android_system_properties",
+ "core-foundation-sys",
+ "js-sys",
+ "wasm-bindgen",
+ "winapi",
+]
  
  [[package]]
  name = "idna"
@@ -275,6 +311,15 @@ dependencies = [
   "cfg-if",
  ]
  
+[[package]]
+name = "js-sys"
+version = "0.3.59"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "258451ab10b34f8af53416d1fdab72c22e805f0c92a1136d59470ec0b11138b2"
+dependencies = [
+ "wasm-bindgen",
+]
+
  [[package]]
  name = "lazy_static"
  version = "1.4.0"
@@ -289,9 +334,9 @@ checksum = "478ee9e62aaeaf5b140bd4138753d1f109765488581444218d3ddda43234f3e8"
  
  [[package]]
  name = "libc"
-version = "0.2.127"
+version = "0.2.131"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "505e71a4706fa491e9b1b55f51b95d4037d0821ee40131190475f692b35b009b"
+checksum = "04c3b4822ccebfa39c02fc03d1534441b22ead323fa0f48bb7ddd8e6ba076a40"
  
  [[package]]
  name = "libnewsboat"
@@ -363,6 +408,15 @@ dependencies = [
   "winapi",
  ]
  
+[[package]]
+name = "log"
+version = "0.4.17"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "abb12e687cfb44aa40f41fc3978ef76448f9b6038cad6aef4259d3c095a2382e"
+dependencies = [
+ "cfg-if",
+]
+
  [[package]]
  name = "malloc_buf"
  version = "0.0.6"
@@ -471,9 +525,9 @@ dependencies = [
  
  [[package]]
  name = "object"
-version = "0.28.4"
+version = "0.29.0"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "e42c982f2d955fac81dd7e1d0e1426a7d702acd9c98d19ab01083a6a0328c424"
+checksum = "21158b2c33aa6d4561f1c0a6ea283ca92bc54802a93b263e910746d679a7eb53"
  dependencies = [
   "memchr",
  ]
@@ -504,9 +558,9 @@ checksum = "eb9f9e6e233e5c4a35559a617bf40a4ec447db2e84c20b55a6f83167b7e57872"
  
  [[package]]
  name = "proc-macro2"
-version = "1.0.40"
+version = "1.0.43"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "dd96a1e8ed2596c337f8eae5f24924ec83f5ad5ab21ea8e455d3566c69fbcaf7"
+checksum = "0a2ca2c61bc9f3d74d2886294ab7b9853abd9c1ad903a3ac7815c58989bb7bab"
  dependencies = [
   "unicode-ident",
  ]
@@ -545,9 +599,9 @@ checksum = "a993555f31e5a609f617c12db6250dedcac1b0a85076912c436e6fc9b2c8e6a3"
  
  [[package]]
  name = "quote"
-version = "1.0.20"
+version = "1.0.21"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "3bcdf212e9776fbcb2d23ab029360416bb1706b1aea2d1a5ba002727cbcab804"
+checksum = "bbe448f377a7d6961e30f5955f9b8d106c3f5e449d493ee1b125c1d43c2b5179"
  dependencies = [
   "proc-macro2",
  ]
@@ -593,9 +647,9 @@ dependencies = [
  
  [[package]]
  name = "redox_syscall"
-version = "0.2.13"
+version = "0.2.16"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "62f25bc4c7e55e0b0b7a1d43fb893f4fa1361d0abe38b9ce4f323c2adfe6ef42"
+checksum = "fb5a58c1855b4b6819d59012155603f0b22ad30cad752600aadfcb695265519a"
  dependencies = [
   "bitflags",
  ]
@@ -613,9 +667,9 @@ dependencies = [
  
  [[package]]
  name = "regex"
-version = "1.5.6"
+version = "1.6.0"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d83f127d94bdbcda4c8cc2e50f6f84f4b611f69c902699ca385a39c3a75f9ff1"
+checksum = "4c4eb3267174b8c6c2f654116623910a0fef09c4753f8dd83db29c48a0df988b"
  dependencies = [
   "aho-corasick",
   "memchr",
@@ -634,9 +688,9 @@ dependencies = [
  
  [[package]]
  name = "regex-syntax"
-version = "0.6.26"
+version = "0.6.27"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "49b3de9ec5dc0a3417da371aab17d729997c15010e7fd24ff707773a33bddb64"
+checksum = "a3f87b73ce11b1619a3c6332f45341e0047173771e8b8b73f87bfeefb7b56244"
  
  [[package]]
  name = "remove_dir_all"
@@ -667,9 +721,9 @@ dependencies = [
  
  [[package]]
  name = "scratch"
-version = "1.0.1"
+version = "1.0.2"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "96311ef4a16462c757bb6a39152c40f58f31cd2602a40fceb937e2bc34e6cbab"
+checksum = "9c8132065adcfd6e02db789d9285a0deb2f3fcb04002865ab67d5fb103533898"
  
  [[package]]
  name = "section_testing"
@@ -686,9 +740,9 @@ dependencies = [
  
  [[package]]
  name = "syn"
-version = "1.0.98"
+version = "1.0.99"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "c50aef8a904de4c23c788f104b7dddc7d6f79c647c7c8ce4cc8f73eb0ca773dd"
+checksum = "58dbef6ec655055e20b86b15a8cc6d439cca19b667537ac6a1369572d151ab13"
  dependencies = [
   "proc-macro2",
   "quote",
@@ -726,18 +780,18 @@ dependencies = [
  
  [[package]]
  name = "thiserror"
-version = "1.0.31"
+version = "1.0.32"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "bd829fe32373d27f76265620b5309d0340cb8550f523c1dda251d6298069069a"
+checksum = "f5f6586b7f764adc0231f4c79be7b920e766bb2f3e51b3661cdb263828f19994"
  dependencies = [
   "thiserror-impl",
  ]
  
  [[package]]
  name = "thiserror-impl"
-version = "1.0.31"
+version = "1.0.32"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "0396bc89e626244658bef819e22d0cc459e795a5ebe878e6ec336d1674a8d79a"
+checksum = "12bafc5b54507e0149cdf1b145a5d80ab80a90bcd9275df43d4fff68460f6c21"
  dependencies = [
   "proc-macro2",
   "quote",
@@ -778,15 +832,15 @@ checksum = "099b7128301d285f79ddd55b9a83d5e6b9e97c92e0ea0daebee7263e932de992"
  
  [[package]]
  name = "unicode-ident"
-version = "1.0.1"
+version = "1.0.3"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5bd2fe26506023ed7b5e1e315add59d6f584c621d037f9368fea9cfb988f368c"
+checksum = "c4f5b37a154999a8f3f98cc23a628d850e154479cd94decf3414696e12e31aaf"
  
  [[package]]
  name = "unicode-normalization"
-version = "0.1.20"
+version = "0.1.21"
  source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "81dee68f85cab8cf68dec42158baf3a79a1cdc065a8b103025965d6ccb7f6cbd"
+checksum = "854cbdc4f7bc6ae19c820d44abdc3277ac3e1b2b93db20a636825d9322fb60e6"
  dependencies = [
   "tinyvec",
  ]
@@ -836,6 +890,60 @@ version = "0.11.0+wasi-snapshot-preview1"
  source = "registry+https://github.com/rust-lang/crates.io-index"
  checksum = "9c8d87e72b64a3b4db28d11ce29237c246188f4f51057d65a7eab63b7987e423"
  
+[[package]]
+name = "wasm-bindgen"
+version = "0.2.82"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "fc7652e3f6c4706c8d9cd54832c4a4ccb9b5336e2c3bd154d5cccfbf1c1f5f7d"
+dependencies = [
+ "cfg-if",
+ "wasm-bindgen-macro",
+]
+
+[[package]]
+name = "wasm-bindgen-backend"
+version = "0.2.82"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "662cd44805586bd52971b9586b1df85cdbbd9112e4ef4d8f41559c334dc6ac3f"
+dependencies = [
+ "bumpalo",
+ "log",
+ "once_cell",
+ "proc-macro2",
+ "quote",
+ "syn",
+ "wasm-bindgen-shared",
+]
+
+[[package]]
+name = "wasm-bindgen-macro"
+version = "0.2.82"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "b260f13d3012071dfb1512849c033b1925038373aea48ced3012c09df952c602"
+dependencies = [
+ "quote",
+ "wasm-bindgen-macro-support",
+]
+
+[[package]]
+name = "wasm-bindgen-macro-support"
+version = "0.2.82"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "5be8e654bdd9b79216c2929ab90721aa82faf65c48cdf08bdc4e7f51357b80da"
+dependencies = [
+ "proc-macro2",
+ "quote",
+ "syn",
+ "wasm-bindgen-backend",
+ "wasm-bindgen-shared",
+]
+
+[[package]]
+name = "wasm-bindgen-shared"
+version = "0.2.82"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "6598dd0bd3c7d51095ff6531a5b23e02acdc81804e30d8f07afb77b7215a140a"
+
  [[package]]
  name = "winapi"
  version = "0.3.9"
```

I don't have lynx intalled.

My guess is that it gets firefox via src/configcontainer.cpp which 
calls utils::get_default_browser which supplies firefox as an option.
